java/ql/lib/semmle/code/java/os/OSCheck.qll

```diff
@@ -0,0 +1,93 @@
+/**
+ * Provides classes and predicates for guards that check for the current OS.
+ */
+
+import java
+import semmle.code.java.controlflow.Guards
+private import semmle.code.java.frameworks.apache.Lang
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.StringCheck
+
+/**
+ * A complimentatry guard that checks if the current platform is Windows.
+ */
+abstract class IsWindowsGuard extends Guard { }
+
+/**
+ * A complimentatry guard that checks if the current platform is unix or unix-like.
+ */
+abstract class IsUnixGuard extends Guard { }
+
+/**
+ * Holds when the MethodAccess is a call to check the current OS using either the upper case `osUpperCase` or lower case `osLowerCase` string constants.
+ */
+bindingset[osUpperCase, osLowerCase]
+private predicate isOsFromSystemProp(MethodAccess ma, string osUpperCase, string osLowerCase) {
+  exists(MethodAccessSystemGetProperty sgpMa |
+    sgpMa.hasCompileTimeConstantGetPropertyName("os.name")
+  |
+    DataFlow::localExprFlow(sgpMa, ma.getQualifier()) and // Call from System.getProperty to some partial match method
```

Comment:
Why the asymmetry between how isUpper and isLowerCase are used?

Reply:
I'm looking at the various ways that people describe to check the OS here:
https://stackoverflow.com/questions/228477/how-do-i-programmatically-determine-operating-system-in-java

Reply:
I think it's safe to assume that if people are using `get("os.name").toLowerCase().startsWith(...)` then someone out there will be using `toUpperCase` too (but see comment above for a proposed fix)